File: backend/src/domain/value_objects/ingestion.py
from dataclasses import dataclass
from enum import Enum
from typing import Any, Optional


class DataSourceType(str, Enum):
    FILE = "FILE"
    MYSQL = "MYSQL"


class FileFormat(str, Enum):
    CSV = "CSV"
    XLSX = "XLSX"
    TXT = "TXT"
    PDF = "PDF"
    DOCX = "DOCX"


class CleaningRuleType(str, Enum):
    NOT_NULL = "NOT_NULL"
    RANGE = "RANGE"
    REGEX = "REGEX"
    DEDUPE = "DEDUPE"


class JobStatus(str, Enum):
    PENDING = "PENDING"
    RUNNING = "RUNNING"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"


class ProcessingMode(str, Enum):
    SYNC = "SYNC"
    ASYNC = "ASYNC"


@dataclass(frozen=True)
class CleaningRule:
    id: str
    field: str
    rule_type: CleaningRuleType
    params: dict[str, Any]
    message: Optional[str] = None


@dataclass(frozen=True)
class FileArtifact:
    artifact_id: str
    project_id: str
    stored_path: str
    file_format: FileFormat
    size_bytes: int
    uploaded_by: str
    checksum: Optional[str] = None

File: frontend/src/types/ingestion.ts
export type DataSourceType = 'FILE' | 'MYSQL'
export type CleaningRuleType = 'NOT_NULL' | 'RANGE' | 'REGEX' | 'DEDUPE'
export type JobStatus = 'PENDING' | 'RUNNING' | 'COMPLETED' | 'FAILED'
export type ProcessingMode = 'SYNC' | 'ASYNC'

export interface CleaningRuleTemplate {
  key: string
  label: string
  description: string
  paramsSchema?: Record<string, unknown>
}

export interface CleaningRule {
  id?: string
  type: CleaningRuleType
  field: string
  params?: Record<string, unknown>
  severity?: 'INFO' | 'WARN' | 'ERROR'
}

export interface MysqlConfig {
  host: string
  port: number
  database: string
  username: string
  password?: string
  table?: string
  ssl?: boolean
}

export interface DataSource {
  id: string
  projectId: string
  name: string
  type: DataSourceType
  status: 'ACTIVE' | 'INACTIVE'
  config: Record<string, unknown>
  createdAt?: string
  lastUsedAt?: string | null
}

export interface DataSourcePayload {
  projectId: string
  type: DataSourceType
  name: string
  mysql?: MysqlConfig
}

export type FileFormat = 'CSV' | 'XLSX' | 'TXT' | 'PDF' | 'DOCX'

export interface UploadArtifact {
  id: string
  projectId: string
  sourceId?: string
  fileFormat: FileFormat
  originalFilename: string
  storedPath: string
  rowCount?: number
  sizeBytes?: number
  createdAt: string
}

export interface IngestionJob {
  id: string
  projectId: string
  artifactId?: string
  sourceId?: string
  mode: ProcessingMode
  status: JobStatus
  progress: number
  totalRows?: number
  processedRows?: number
  errorMessage?: string | null
  resultPath?: string | null
  createdAt?: string
  updatedAt?: string
}

export interface FileUploadResponse {
  artifactId: string
  jobId?: string | null
  mode: ProcessingMode
  status: JobStatus
  previewRows?: Array<Record<string, unknown>>
}

export interface FileUploadPayload {
  projectId: string
  sourceId?: string
  file: File
  rules?: CleaningRule[]
}

export interface PreviewPayload {
  projectId: string
  artifactId: string
  rules: CleaningRule[]
}

export interface MysqlImportPayload {
  projectId: string
  sourceId: string
  table?: string
  rules?: CleaningRule[]
}

export interface MysqlTestResult {
  ok: boolean
  message: string
  sampleRows?: Array<Record<string, unknown>>
}

export interface WizardFormState {
  projectId: string
  sourceType: DataSourceType
  selectedSourceId?: string
  mysql: MysqlConfig
  cleaningRules: CleaningRule[]
}
